### PR TITLE
[3.3.5] Quick script re-write: "Mirror Image" (Mage spell)

### PR DESCRIPTION
**Pull request by Treeston**
Current one is a colossal clusterfuck of grid searchers all over the place, which I stumbled into by chance. Rather than deal with the current one, I decided to streamline and rewrite.

### CODE REVIEW

File: src/server/scripts/Pet/pet_mage.cpp [hunk 1]
```diff
@@ -40,196 +40,150 @@ enum MageSpells
...
-                me->VisitNearbyObject(40.0f, searcher);
-
-                for (std::list<Unit*>::const_iterator iter = targets.begin(); iter != targets.end(); ++iter)
+                
+                if (!selectedTarget) // no pvp targets found, look for pve combat targets instead
                 {
-                    if ((*iter) == target)
+                    float threatThreshold = 0.0f;
+                    if (selectedTarget)
+                        threatThreshold = selectedTarget->getThreatManager().getThreat(owner) * 1.1f;
```
Comment (ccrs) [inline src/server/scripts/Pet/pet_mage.cpp]: cant get in

Comment (Treeston) [inline src/server/scripts/Pet/pet_mage.cpp]: ?

**Comment by ccrs** [comment]
umm they dont attack the dummies, consider owners current selected target?

File: src/server/scripts/Pet/pet_mage.cpp (hunk 1 above)
Comment (ccrs) [inline src/server/scripts/Pet/pet_mage.cpp]:
````
if (!selectedTarget)
````

Comment (Treeston) [inline src/server/scripts/Pet/pet_mage.cpp]: Damn, good point.

Comment (Treeston) [inline src/server/scripts/Pet/pet_mage.cpp]: Fxied

**Comment by Treeston** [comment]
Dummy behavior fixed too.

File: src/server/game/Entities/Unit/Unit.cpp [hunk 2]
```diff
@@ -5862,6 +5862,15 @@ bool Unit::Attack(Unit* victim, bool meleeAttack)
         if (victim->GetTypeId() == TYPEID_PLAYER)
             victim->SetInCombatWith(this);
 
+        // also enter combat with pet/guardian's owner if exists
+        if (Unit* owner = victim->GetOwner())
```
Comment (ccrs) [inline src/server/game/Entities/Unit/Unit.cpp]: so... If I send my pet/guardian to attack a creature I will be in combat?

Comment (Treeston) [inline src/server/game/Entities/Unit/Unit.cpp]: Other way around. If a creature attacks your pet/guardian, you will be in combat.

Comment (ccrs) [inline src/server/game/Entities/Unit/Unit.cpp]: what about a pvp combat where a warlock pet attacks a hunter pet? both owners shouldnt be in combat

Comment (ccrs) [inline src/server/game/Entities/Unit/Unit.cpp]: nvm that, the check above will prevent it

**Comment by Treeston** [comment]
This will be delayed until #19930 is merged.

**Comment by DanVS** [comment]
#19930 was merged. Any news on this?

**Comment by tkrokli** [comment]
Combat/Threat system rewrite caused a few crash issues and maybe some more bugs. Fixing those is higher priority, so it will cause a slight delay after the merge of #19930 .

**Comment by Treeston** [comment]
New version is already pushed to the branch (99d9afb, just above) and ready for testing.

**Comment by DanVS** [comment]
I have found that mirrored images doesn't follow mage when combat finishes.

**Pull request merged**